utils/big_decimal: add fast paths to operator <=>

denesb opened the pull request:
Currently, the tri-compare operator for big_decimal (operator <=>), uses a precise but potentially very expensive algorithm for comparing the numbers: it first brings them to the same scale, then compares the normalized unscaled values. big_decimal has abritrary precisions, therefore the stored numbers can be arbitrarily large. In extreme cases, comparing two numbers can result in huge amount of memory allocated and stalls. If this type is used int he primary key of a table, these comparisons can make the node completely unresponsive. 

This patch adds the following fast-paths to operator <=>:
* An early return for the case of equal scales.
* An early return for different signs.
* An early return for the case where one or both of the numbers are 0.
* A fast algorithm for detecting the case where the there is a big difference between the two numbers. This algorithm works only with the scales and is able to compare the two numbers by using only one division and some additions and substractions. This algorithm is imprecise and when the numbers are closer than its confidence window, it will fall-back to the current slow but precise tri-compare.

All but the last case should have been fast before as well, but the scale-compare algorithm makes a huge difference. Numbers, which would previously make the node unresponsive, now compare in constant-time.

Fixes: https://github.com/scylladb/scylladb/issues/21716

avikivity commented:
tests!!!!!!eleven

avikivity commented on utils/big_decimal.cc:
```diff
@@ -143,6 +143,35 @@ std::strong_ordering big_decimal::operator<=>(const big_decimal& other) const
 
+std::strong_ordering big_decimal::operator<=>(const big_decimal& other) const
+{
+    if (_scale == other._scale) {
```
Even before: compare the sign function on the two mantissas, if they are different, or if both are zero, that's the result.

denesb replied:
Yes, I forgot about the sign.

denesb replied:
Done in v2.

avikivity commented on utils/big_decimal.cc:
```diff
@@ -143,6 +143,35 @@ std::strong_ordering big_decimal::operator<=>(const big_decimal& other) const
+    const static double sqrt_10 = std::sqrt(10);
+
+    const int64_t bit_count = boost::multiprecision::msb(_unscaled_value) - boost::multiprecision::lsb(_unscaled_value);
+    const int64_t bit_count_other = boost::multiprecision::msb(other._unscaled_value) - boost::multiprecision::lsb(other._unscaled_value);
```
What does msb mean on a negative mantissa?

avikivity replied:
Not sure why you care about the lsb. It's basically random.

nyh replied:
I don't understand this algorithm at all, so you definitely need a unit test to test it.

denesb replied:
I added detailed comments to the code, as well as extensive tests in v2.

denesb replied:
(I also dropped the unnecessary use of `lsb()` in v2).

avikivity commented on utils/big_decimal.cc:
```diff
@@ -143,6 +143,35 @@ std::strong_ordering big_decimal::operator<=>(const big_decimal& other) const
+
+    const int64_t diff_scale_2 = bit_count - bit_count_other;
+    const double diff_scale_2_adj_10 = double(diff_scale_2) / sqrt_10;
+    const double diff_scale_10 = _scale - other._scale;
```
Don't understand sqrt(10) here.

I'd expect logarithms. v = unscaled * 10^scale. Then log10 v = log10(unscaled) + scale =~ log2(unscaled)/log2(10) + scale. log2(unscaled) can be approximated with msb(abs(unscaled)), but I don't see sqrt(10) anywhere.

denesb replied:
Yes, I should have used `log2(10)`, this is fixed in v2.

michoecho reviewed (changes requested):
The general approach (you are trying to compare $a \cdot 10^{-A}$ with $b \cdot 10^{-B}$ by computing

$$\log_{10} \frac{a \cdot 10^{-A}}{b \cdot 10^{-B}} = \log_{10} {a} - \log_{10} {b} - A + B = \frac{\log_{2} {a}}{\log_{2} {10}} - \frac{\log_{2} {b}}{\log_{2} {10}} - A + B$$

inexactly and checking that both sides of the confidence interval have the same sign) is sound, but:

1. You forgot that `_unscaled_value` can be negative or 0. In this case logarithms don't work. You have to deal with the signs separately, and only apply the logarithms to positive values. (And, appropriately, boost will throw an exception if you apply `msb` to something non-positive anyway).
2. `lsb` doesn't influence the logarithm, you only care about `msb` (because $\log_{2}{x} = \\texttt{msb}(x) + \epsilon | \epsilon \in {(0, 1]})$.
3. You meant $\log_2{10}$ but wrote $\texttt{sqrt}(10)$.

denesb commented:
I see `lsb` and `msb` is creating a lot of confusion. I don't care about the bits themselves, I care about their index. I fact I only care about bit count in both numbers, and as far as I can tell based on my search on the internet, this is the best way boost got to give it to you...

avikivity commented:
> I see `lsb` and `msb` is creating a lot of confusion. I don't care about the bits themselves, I care about their index. I fact I only care about bit count in both numbers, and as far as I can tell based on my search on the internet, this is the best way boost got to give it to you...

msb is a fine approximation (as long as you remember its limitations) for log2. lsb is just a random number. 1048576 and 1048577 are very close, one has an lsb of 20 and the other an lsb of 2. What did the lsb tell you?

michoecho commented:
> I see `lsb` and `msb` is creating a lot of confusion. I don't care about the bits themselves, I care about their index. I fact I only care about bit count in both numbers, and as far as I can tell based on my search on the internet, this is the best way boost got to give it to you...

No, `msb` is perfectly fine. It's equal to `$\log_{2}{x} = \\texttt{msb}(x) + \epsilon$` which is exactly what we want.

But `lsb` is just nonsense there. Numbers `0x10001` and `0x10000` have the same scale (binary logarithm) of 16. But you are saying that the former has scale `16` while the latter has scale `0`.

denesb commented:
> tests!!!!!!eleven

We already have some, but I will add more, with numbers with big differences, to target the new code.

avikivity commented:
And with negatives and zeros, and one that crashes the thing before the fix

denesb commented:
> > I see `lsb` and `msb` is creating a lot of confusion. I don't care about the bits themselves, I care about their index. I fact I only care about bit count in both numbers, and as far as I can tell based on my search on the internet, this is the best way boost got to give it to you...
> 
> No, `msb` is perfectly fine. It's equal to log 2 ⁡ x = msb ( x ) + ϵ [ 0 ; 1 ) which is exactly what we want.
> 
> But `lsb` is just nonsense there. Numbers `0x10001` and `0x10000` have the same scale (i.e. binary logarithm) equal to 16 + ϵ . But you are saying that the former has scale `16` while the latter has scale `0`.

I feel like I also misunderstood what msb and lsb is. I *thought* that `msb` and `lsb` is the index of the most significant and least significant bit respectively, in the vector of bits representing the number. This vector can be like circular_buffer, i.e. the elements don't stat from 0.

But according to the above, my understanding is wrong.

michoecho commented:
> > > I see `lsb` and `msb` is creating a lot of confusion. I don't care about the bits themselves, I care about their index. I fact I only care about bit count in both numbers, and as far as I can tell based on my search on the internet, this is the best way boost got to give it to you...
> > 
> > 
> > No, `msb` is perfectly fine. It's equal to log 2 ⁡ x = msb ( x ) + ϵ [ 0 ; 1 ) which is exactly what we want.
> > But `lsb` is just nonsense there. Numbers `0x10001` and `0x10000` have the same scale (i.e. binary logarithm) equal to 16 + ϵ . But you are saying that the former has scale `16` while the latter has scale `0`.
> 
> I feel like I also misunderstood what msb and lsb is. I _thought_ that `msb` and `lsb` is the index of the most significant and least significant bit respectively, in the vector of bits representing the number. This vector can be like circular_buffer, i.e. the elements don't stat from 0.
> 
> But according to the above, my understanding is wrong.

Yep, your understanding was wrong. `lsb` is the index of the least significant `1` bit.

denesb commented:
> > > > I see `lsb` and `msb` is creating a lot of confusion. I don't care about the bits themselves, I care about their index. I fact I only care about bit count in both numbers, and as far as I can tell based on my search on the internet, this is the best way boost got to give it to you...
> > > 
> > > 
> > > No, `msb` is perfectly fine. It's equal to log 2 ⁡ x = msb ( x ) + ϵ [ 0 ; 1 ) which is exactly what we want.
> > > But `lsb` is just nonsense there. Numbers `0x10001` and `0x10000` have the same scale (i.e. binary logarithm) equal to 16 + ϵ . But you are saying that the former has scale `16` while the latter has scale `0`.
> > 
> > 
> > I feel like I also misunderstood what msb and lsb is. I _thought_ that `msb` and `lsb` is the index of the most significant and least significant bit respectively, in the vector of bits representing the number. This vector can be like circular_buffer, i.e. the elements don't stat from 0.
> > But according to the above, my understanding is wrong.
> 
> Yep, your understanding was wrong. `lsb` is the index of the least significant `1` bit.

Ah, now I understand why you think `lsb` is useless. Indeed it is, and we only need `msb` to determine the number of bits that make up the number.

avikivity commented:
> > > I see `lsb` and `msb` is creating a lot of confusion. I don't care about the bits themselves, I care about their index. I fact I only care about bit count in both numbers, and as far as I can tell based on my search on the internet, this is the best way boost got to give it to you...
> > 
> > 
> > No, `msb` is perfectly fine. It's equal to log 2 ⁡ x = msb ( x ) + ϵ [ 0 ; 1 ) which is exactly what we want.
> > But `lsb` is just nonsense there. Numbers `0x10001` and `0x10000` have the same scale (i.e. binary logarithm) equal to 16 + ϵ . But you are saying that the former has scale `16` while the latter has scale `0`.
> 
> I feel like I also misunderstood what msb and lsb is. I _thought_ that `msb` and `lsb` is the index of the most significant and least significant bit respectively, in the vector of bits representing the number. This vector can be like circular_buffer, i.e. the elements don't stat from 0.


most significant set bit and least significant set bit.


> 
> But according to the above, my understanding is wrong.

Not wrong, but lsb is just not useful for anything.

denesb commented:
There are two questions related to the use of sqrt(10):
* https://github.com/scylladb/scylladb/pull/21715#discussion_r1862065435 from @avikivity 
* https://github.com/scylladb/scylladb/pull/21715#pullrequestreview-2468022310 from @michoecho 

I'll try to answer both here, by describing the algorithm:
```
// We have two numbers in the form:
auto X = unscaled_x * (10 ** scale_x);
auto Y = unscaled_y * (10 ** scale_y);

// The idea behind the algorithm is to bring these to a form where I can compare just the scales.
auto N = unscaled_x / unscaled_y;

// Now we can rewrite the original numbers as this:
auto X1 = unscaled_x * (10 ** scale_x);
auto Y1 = unscaled_x * N * (10 ** scale_y);

// These are equivalent
assert(X1 == X);
assert(Y1 == Y);

// Work on Y1
auto M = log2(N);
auto Y2 = unscaled_x * (2**M) * (10**scale_y);

// Y2 != Y, but it is within a factor of 2
assert(Y2 > Y/2 && Y2 < Y*2);

// Now Y2 and X1 have the same unscaled bases, so we can omit them when comparing
// We still have different power bases though
// We can convert 2**a -> 10**b, where b = a/sqrt(10)
auto L = N/sqrt(10);

// Now the scales are directly comparable
auto Y3 = unscaled_x * (10**L) * (10**scale_y);
Y3 = unscaled_x * (10**(L + scale_y));
```

nyh commented:
@denesb you need a regression test, preferably a cqlpy test which fails (OOMs, crashes, errors, etc.) before your patch, and passes afterwards.

It would also be good to have a unit test (in C++, but can also be cqlpy if you want) to do all sorts of delicate comparisons that you discovered while developing this code. For example, you said you have a special case if the two numbers are "close" - the test should try such cases and make sure that these cases work - and that they don't regress later.

michoecho commented:
> ```
> // We can convert 2**a -> 10**b, where b = a/sqrt(10)
> ```

@denesb The thing is: if you want a `b` such that `2**a == 10**b`, then you must take `b := a / log2(10)`, not `b := a / sqrt(10)`. That's what we were telling you about the `sqrt`.

denesb commented:
> > There are two questions related to the use of sqrt(10):
> > ...
> > ```
> > // We can convert 2**a -> 10**b, where b = a/sqrt(10)
> > ```
> 
> @denesb The thing is: if you want a `b` such that `2**a == 10**b`, then you must take `b := a / log2(10)`, not `b := a / sqrt(10)`. That's what we were telling you about the `sqrt`.

Ah, yes. You are right of course, I mixed up sqrt with log2.

denesb commented:
> > > There are two questions related to the use of sqrt(10):
> > > ...
> > > ```
> > > // We can convert 2**a -> 10**b, where b = a/sqrt(10)
> > > ```
> > 
> > 
> > @denesb The thing is: if you want a `b` such that `2**a == 10**b`, then you must take `b := a / log2(10)`, not `b := a / sqrt(10)`. That's what we were telling you about the `sqrt`.
> 
> Ah, yes. You are right of course, I mixed up sqrt with log2.

Unfortunately, log2(10) is very close to sqrt(10) so I didn't notice this error when doing my test calculations.

denesb commented:
> > I'll try to answer both here, by describing the algorithm:
> 
> Please explain this algorithm in a code comment - nobody will be able to guess why this code does what it does without a clear explanation.

I will inline the above explanation into the code, so the algorithm is explained as the code goes along.

> And of course, if your tests (where are they?) didn't see that sqrt(10) was wrong, write another test which is more sensitive to the fact that sqrt(10) != log2(10).

I don't have yet, I should have made this an RFC, I published my code to get early feedback.

> 
> Please also consider mentioning explicitly the issue that the "slow" algorithm isn't just slow - its most serious problem is that it needs to rescale both numbers to the same scale which can cause a huge allocation. Your improvement simply recognizes quickly that the two numbers are very different magnitude, so you can immediately tell which is higher without copying anything.

I described the problem in the issue in detail, even included a backtrace. I will add some more detauils to the patch too.

michoecho commented:
> > > > There are two questions related to the use of sqrt(10):
> > > > ...
> > > > ```
> > > > // We can convert 2**a -> 10**b, where b = a/sqrt(10)
> > > > ```
> > > 
> > > 
> > > @denesb The thing is: if you want a `b` such that `2**a == 10**b`, then you must take `b := a / log2(10)`, not `b := a / sqrt(10)`. That's what we were telling you about the `sqrt`.
> > 
> > 
> > Ah, yes. You are right of course, I mixed up sqrt with log2.
> 
> Unfortunately, log2(10) is very close to sqrt(10) so I didn't notice this error when doing my test calculations.

@denesb If $\log_2(10)$ is "very close" to $\sqrt{10}$ then  $\sqrt{10}$ is practically equal to $\pi$ :stuck_out_tongue:

> > I'll try to answer both here, by describing the algorithm:
> 
> Please explain this algorithm in a code comment - nobody will be able to guess why this code does what it does without a clear explanation.

@nyh While I agree that this is tricky enough to warrant more comments, I claim it's still more intuitive than 90% of Scylla's codebase.

> And of course, if your tests (where are they?) didn't see that sqrt(10) was wrong, write another test which is more sensitive to the fact that sqrt(10) != log2(10).

Tests are a bad tool for this kind of problem. A test designed specifically to differentiate between a formula with  `log2(10)` and a formula with `sqrt(10)` is too specific. It's useless for anything except ensuring that someone doesn't switch the constant back to `sqrt(10)`. A test that tests all possible inputs is infeasible. Even a test that tests all possible inputs up to a certain size might be too computationally expensive to differentiate those two constants in a reasonable time. So how do you even approach designing a reasonable test here?

nyh commented:
> > Please also consider mentioning explicitly the issue that the "slow" algorithm isn't just slow - its most serious problem is that it needs to rescale both numbers to the same scale which can cause a huge allocation. Your improvement simply recognizes quickly that the two numbers are very different magnitude, so you can immediately tell which is higher without copying anything.
> 
> I described the problem in the issue in detail, even included a backtrace. I will add some more detauils to the patch too.

Yes, I meant explain it in the code comments - where currently you focused only on the adjective "slow", which isn't exactly the main problem you are solving.

I think this is one of the cases where a big comment explaining the entire algorithm (and its purpose) is more helpful than small comments before each line.

nyh commented:
> > Please explain this algorithm in a code comment - nobody will be able to guess why this code does what it does without a clear explanation.
> 
> @nyh While I agree that this is tricky enough to warrant more comments, I claim it's still more intuitive than 90% of Scylla's codebase.

I guess 90% of Scylla's codebase doesn't have enough comments as well.

But seriously, what I was looking for isn't comments like:
```c++
// multiply x by log2(10)
x *= log2(10)
```
But a high-level description of the purpose of the algorithm and what it does - similar to what Botond write in this thread but not in the code.


> > And of course, if your tests (where are they?) didn't see that sqrt(10) was wrong, write another test which is more sensitive to the fact that sqrt(10) != log2(10).
> 
> Tests are a bad tool for this kind of problem. A test designed specifically to differentiate between a formula with `log2(10)` and a formula with `sqrt(10)` is too specific.

Sure, I know. But now that he knows his first implementation had a specific bug, he can add a test case that convinces him that it really was a bug.
He can also write a test that tries in a loop all sorts of numbers (that's why I like tests with a real programming language, not "approval tests" like test/cql) and this may have caught this bug (although it's possible it won't).

> It's useless for anything except ensuring that someone doesn't switch the constant back to `sqrt(10)`.

That's true. But it also helps Botond see for himself that no, sqrt(10) was wrong - if it wasn't already obvious to him.
In any case, I'm not saying this is the most important test to write. Out of all the tests that are needed for this feature, I agree that is the least important one.

denesb commented:
> > Tests are a bad tool for this kind of problem. A test designed specifically to differentiate between a formula with `log2(10)` and a formula with `sqrt(10)` is too specific.
> 
> Sure, I know. But now that he knows his first implementation had a specific bug, he can add a test case that convinces him that it really was a bug.

I think *this* specific test is not possible to write, because of how large I made the no-confidence window. I can make it narrower if somebody can make a good case for a lower window (with proof). I lack confidence in my math abilities to do that.

michoecho commented:
> I can make it narrower if somebody can make a good case for a lower window (with proof). I lack confidence in my math abilities to do that.

@denesb 

Let:

```math
\begin{aligned}
\texttt{a} &:= \texttt{left\_bd.\_unscaled\_value} \\
\texttt{A} &:= \texttt{left\_bd.\_scale} \\
\texttt{b} &:= \texttt{right\_bd.\_unscaled\_value} \\
\texttt{B} &:= \texttt{right\_bd.\_scale} \\
\end{aligned}
```
Then:
```math
\begin{aligned}
\texttt{left\_bd} &= a \cdot 10^{-A} \\
\texttt{right\_bd} &= b \cdot 10^{-B} \\
\end{aligned}
```

Your algorithm is based on the following observation: if `left_bd` and `right_bd` are positive, then:

```math
\begin{aligned}
& \texttt{left\_bd <=> right\_bd} \\
&= \log_{10}(\texttt{left\_bd}) <=>\log_{10}(\texttt{right\_bd}) \\
&= \log_{10}(a \cdot 10^{-A}) <=>\log_{10}(b \cdot 10^{-B}) \\
&= \log_{10}(a \cdot 10^{-A}) - \log_{10}(b \cdot 10^{-B}) <=> 0 \\
&= \log_{10} {a} - \log_{10} {b} + (B - A) <=> 0 \\
&= \frac{\log_{2} {a} - \log_{2} {b}}{\log_{2} {10}} + (B - A) <=> 0 \\
&= \frac{\texttt{msb}(a) - \texttt{msb}(b)}{\log_{2} {10}} + (B - A) <=> \frac{(\texttt{msb}(a) - \log_{2} {a}) - (\texttt{msb}(b) - \log_{2} {b})}{\log_{2} {10}}
\end{aligned}
```

Now, the left side of the last `<=>` is what your code computes as `diff_scale`. Your goal is to check if `diff_scale` is greater (or smaller) than every possible value of the right side.

But $\texttt{msb}(a) - \log_{2} {a}$ is in range $[0; 1)$ and $\texttt{msb}(b) - \log_{2} {b}$ is in range $[0; 1)$. So the numerator is in range $(-1; 1)$, and the entire fraction is in range $(-\log_{10} 2; \log_{10} 2) \approx (-0.301; 0.301)$. That's the smallest possible "confidence interval", before adding a margin for floating-point approximation errors.

Rounding errors are a thing, but since `_unscaled_value` can be assumed much smaller than $2^{40}$, the total error is going to be off by at most... what, 10 ulps at $2^{40}$, so less than 10 * $2 ^ {-12}$, so less than 0.0025?

So you can narrow it down from $(-2; 2)$ to at least $(-0.31; 0.31)$. But I wouldn't do that.

> I think this specific test is not possible to write, because of how large I made the no-confidence window.

@denesb Not at all. Using `sqrt` instead of `log2` there makes `diff_scale_2_adj_10` greater by a *factor* of `1.05`. So the formula is pretty easy to break. You just need to pick any numbers such that `diff_scale_2_adj_10 > diff_scale_10 + 2`, but `diff_scale_2_adj_10 / 1.05 < diff_scale_10 - 2`. Then the broken formula will answer "greater", while the true answer is "smaller".

michoecho commented:
@denesb One more thing: you can't do `_scale - other._scale`, because this can wrap around. You need to cast the scales to `int64_t` before the subtraction.

denesb commented:
> So you can narrow it down from ( − 2 ; 2 ) to at least ( − 0.31 ; 0.31 ) . But I wouldn't do that.

Thanks for the detailed proof. I reduced it to (-1; 1), which is still ample.

> @denesb Not at all. Using `sqrt` instead of `log2` there makes `diff_scale_2_adj_10` greater by a _factor_ of `1.05`. So the formula is pretty easy to break. You just need to pick any numbers such that `diff_scale_2_adj_10 > -diff_scale_10 + 2`, but `diff_scale_2_adj_10 / 1.05 < -diff_scale_10 - 2`. Then the broken formula will answer "greater", while the true answer is "smaller".

Because the difference between the real confidence window and the one used by the code is larger than this factor of `1.05`, I still think this test is not possible to write.

denesb commented:
New in v2:
* Added tests. Turns out there were no comparison tests at all. The new tests found the following bugs:
    - I wrongly assumed that scale has a positive sign, in fact it has implicit negative sign.
    - Calculating the difference between the diffs can wrap around, this was pointed out by @michoecho [above](https://github.com/scylladb/scylladb/pull/21715#issuecomment-2506749781).
    - Additionally, I added test cases which reproduce the increased CPU usage. I didn't ahve the patiente to wait out the test finishing (doing a single comparison), I killed it after a few dozen seconds.
* Bugfixes (see above).
* Added explicit early returns for mismatching sign and one or both numbers being 0.
* Added detailed explanation to the algorithm in comments.
* Expanded the patchlog with better explanation of the problem.

nyh commented:
> @denesb you need a regression test, preferably a cqlpy test which fails (OOMs, crashes, errors, etc.) before your patch, and passes afterwards.

While the C++ unit tests that you added are important, I still think a CQL test is important too (I personally prefer cqlpy but test/cql and test/boost are also fine). Why? Not because I prefer Python tests ;-) But rather, because your whole fix assumes (and we all trust you, but still...) that the code path actually uses `operator<=>` which you fixed and some other comparison functions. What if some future refactoring of the code changes the code to use some other function like `operator-` or `cmp()` who knows what, and that other function does not have the fix? Moreover, your fix assumes that `operator<=>` was the only problem that broke the user's workload. Is it possible that the code does something else with those decimals, not only `operator<=>`, and that breaks too? You can't know this from a unit test that only checks `operator<=>`.

If you need help, I'll volunteer to write a cqlpy test for this.

nyh commented:
Here is a reproducer, in a new test file `test/cqlpy/test_type_decimal.py`. I tested the problem before your patch, not your patch. The test runs for minutes spewing out warnings of larger and larger allocations and stalls, until the driver gives up getting a response (I didn't see it actually run out of memory, maybe setting a lower memory size of Scylla will make this happen, but it doesn't matter). The test passes on Cassandra.
```python
# Copyright 2024-present ScyllaDB
#
# SPDX-License-Identifier: AGPL-3.0-or-later

#############################################################################
# Test involving the "decimal" column type.
# There are additional tests involving decimals in specific contexts in other
# files - e.g., aggregating decimals in test_aggregate.py, casting decimals
# in test_cast_data.py, and decimals in JSON in test_json.py
#############################################################################

from decimal import Decimal
import pytest

import nodetool
from util import new_test_table, unique_key_int

@pytest.fixture(scope="module")
def table1(cql, test_keyspace):
    with new_test_table(cql, test_keyspace, "p int, c decimal, PRIMARY KEY (p, c)") as table:
        yield table

# Test that if we have clustering keys of wildly different scales, sorting
# them works. Reproduces issue #21716 where comparing two decimals of
# very different scales needed to expand the digits of one of them and
# would take huge amounts of CPU and also run out of memory.
# See also test_json.py::test_json_decimal_high_mantissa where printing
# a decimal runs into a similar problem.
def test_decimal_clustering_key_high_mantissa(cql, table1):
    # The two numbers "low" and "high" have wildly different mantissas.
    # Any algorithm that attempts to expand the digits of one to have the
    # same mantissa as the other will run out of memory (#21716).
    low = Decimal('1e-1000000000')
    high = Decimal('1e1000000000')
    minuslow = Decimal('-1e-1000000000')
    minushigh = Decimal('-1e1000000000')
    p = unique_key_int()
    stmt = cql.prepare(f"INSERT INTO {table1} (p, c) VALUES ({p}, ?)")
    cql.execute(stmt, [high])
    cql.execute(stmt, [low])
    cql.execute(stmt, [minushigh])
    cql.execute(stmt, [minuslow])
    assert list(cql.execute(f"SELECT c from {table1} where p = {p}")) == [(minushigh,),(minuslow,),(low,),(high,)]
    # Do a memtable flush, just to make sure the memtable-flushing or sstable
    # writing code also doesn't use problematic decimal-handling algorithms.
    nodetool.flush(cql, table1)
```

kostja commented:
In ANSI DECIMAL data type precision/scale is fixed at schema definition time. So "scaling" it is unnecessary in the first place. In Java, DECIMAL has arbitrary precision. Please note that tutorials recommend rounding the value before storing it in the database. It is not impossible this wasn't done in the client's case, thus giving us indeed arbitrary scales, which bear no value to the user.

Here's Java implementation of compareTo:

https://github.com/frohoff/jdk8u-dev-jdk/blob/master/src/share/classes/java/math/BigDecimal.java#L2623

kostja commented:
@xemul you requested my review of this patch. Not coincidentally I gave a reference to the Java implementation. I think our safest bet here is to copy Cassandra, which is most likely using java.math.BigDecimal.compareTo(), and not use our own logic.
cc @denesb

raphaelsc commented:
> > In ANSI DECIMAL data type precision/scale is fixed at schema definition time. So "scaling" it is unnecessary in the first place. In Java, DECIMAL has arbitrary precision. Please note that tutorials recommend rounding the value before storing it in the database. It is not impossible this wasn't done in the client's case, thus giving us indeed arbitrary scales, which bear no value to the user.
> 
> I agree that ridiculously high exponents have no common use cases, so nobody will shed tears if they are limited. In the DynamoDB API, the exponent is limited (the decimal exponent can be up to around 130, see #6794), and nobody ever complained (as far as I saw). The problem is that Cassandra - which as you said is based on the Java implementation - is **not** limited. As the test I wrote above shows, `1e1000000000` is accepted just fine in Cassandra. We can decide that Scylla will differ from Cassandra in this point and document this difference. If we give a clear error in this case, user's will know when they reached limitation (and we can make the specific limitation configurable, by the way).
> 
> Perhaps we should because this is not the only Scylla-crashing bug we have in this area - there's also #8002 - we should indeed have (configurable) limiting of the exponent, which will avoid all these bugs in one fell swoop.

That's what I suggested. Spark does set a limit https://spark.apache.org/docs/3.5.1/api/java/org/apache/spark/sql/types/DecimalType.html

We have our own guardrails that aren't present in C*, and I think we could add one for this as well. But it's not mutually exclusive with this patch (that avoids expansion), and we need it to clear the fire on existing installations.

michoecho commented:
> > So you can narrow it down from ( − 2 ; 2 ) to at least ( − 0.31 ; 0.31 ) . But I wouldn't do that.
> 
> Thanks for the detailed proof. I reduced it to (-1; 1), which is still ample.
> 
> > @denesb Not at all. Using `sqrt` instead of `log2` there makes `diff_scale_2_adj_10` greater by a _factor_ of `1.05`. So the formula is pretty easy to break. You just need to pick any numbers such that `diff_scale_2_adj_10 > -diff_scale_10 + 2`, but `diff_scale_2_adj_10 / 1.05 < -diff_scale_10 - 2`. Then the broken formula will answer "greater", while the true answer is "smaller".
> 
> Because the difference between the real confidence window and the one used by the code is larger than this factor of `1.05`, I still think this test is not possible to write.

@denesb The thing is, the factor isn't applied to `diff_scale`, but to `diff_scale_2_adj_10`. For *any* finite confidence interval K you can find arguments such that `diff_scale_2_adj_10 - diff_scale_10` is on one side of K, but `diff_scale_2_adj_10 * 1.05 - diff_scale_10` is on the other side of K. You just need to make `diff_scale_2_adj_10 * 0.05` bigger than the width of the interval, and then use `diff_scale_10` to shift the two possible results to the right place.

So, for example, for the interval (-2, 2), a comparison of $1$ with $2^{400} \cdot 10^{123}$ should do the trick, because the good version will compute `diff_scale = 400 / log2(10) - 123 = ~-2.5`, while the bad version will compute `diff_scale = 400 / sqrt(10) - 123 = ~+3.5`

> New in v2:

@denesb The logarithm-based code path still doesn't work for negative `_unscaled_value`. (None of the tests covers this case). Before entering that path, you check that both numbers have the same sign, but that's not enough. If both numbers are negative, then you have to invert the result of the magnitude comparison at the end. (And you somehow to negate the argument to `msb` because boost will throw if you pass a negative number there).

michoecho commented:
> @xemul you requested my review of this patch. Not coincidentally I gave a reference to the Java implementation. I think our safest bet here is to copy Cassandra, which is most likely using java.math.BigDecimal.compareTo() (needs to be checked), and not use our own logic. cc @denesb

@kostja You mean the semantics, or the implementation? As far as I can see in the source you linked, the Java implementation is prone to the very problem this PR is trying to fix.

nyh commented:
> @kostja You mean the semantics, or the implementation? As far as I can see in the source you linked, the Java implementation is afflicted with the very problem this PR is trying to fix.

As I showed in the cqlpy test above, Cassandra does **not** have the user-facing bug that we have. Whether it uses the specific Java function you mentioned or something else to compare "decimal" clustering keys, I don't know.

kbr-scylla commented:
Regarding tests, this looks like a good candidate for property testing. Generate two numbers, compare them using both the smart algorithm and a dumb-but-obvious algorithm, assert the result is the same. (Generating two numbers in a way that makes the test valuable is the hard part here, e.g. just randomly generating two small numbers won't provide much value)

kostja commented:
> > @xemul you requested my review of this patch. Not coincidentally I gave a reference to the Java implementation. I think our safest bet here is to copy Cassandra, which is most likely using java.math.BigDecimal.compareTo() (needs to be checked), and not use our own logic. cc @denesb
> 
> @kostja You mean the semantics, or the implementation? As far as I can see in the source you linked, the Java implementation is afflicted with the very problem this PR is trying to fix.

How did you arrive to this? AFAICS compareTo avoids inflation as much as it can.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13497/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13497/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/big_decimal_test
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/439/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5066/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13082/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/1958/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13497/)
#### Failed Tests (37/36012):
- [test_number_magnitude_key.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13497/testReport/junit/alternator/test_number/Tests___Unit_Tests___test_number_magnitude_key_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_number_magnitude_key.debug.1)
- [test_number_magnitude_key.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13497/testReport/junit/alternator/test_number/Tests___Unit_Tests___test_number_magnitude_key_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_number_magnitude_key.debug.2)
- [test_number_magnitude_key.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13497/testReport/junit/alternator/test_number/Tests___Unit_Tests___test_number_magnitude_key_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_number_magnitude_key.debug.3)
- [test_query_sort_order_number.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13497/testReport/junit/alternator/test_query/Tests___Unit_Tests___test_query_sort_order_number_debug_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_query_sort_order_number.debug.1)
- [test_query_sort_order_number.debug.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13497/testReport/junit/alternator/test_query/Tests___Unit_Tests___test_query_sort_order_number_debug_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_query_sort_order_number.debug.2)
- [test_query_sort_order_number.debug.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13497/testReport/junit/alternator/test_query/Tests___Unit_Tests___test_query_sort_order_number_debug_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_query_sort_order_number.debug.3)
- [test_number_magnitude_key.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13497/testReport/junit/alternator/test_number/Tests___Unit_Tests___test_number_magnitude_key_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_number_magnitude_key.dev.1)
- [test_number_magnitude_key.dev.2](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13497/testReport/junit/alternator/test_number/Tests___Unit_Tests___test_number_magnitude_key_dev_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_number_magnitude_key.dev.2)
- [test_number_magnitude_key.dev.3](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13497/testReport/junit/alternator/test_number/Tests___Unit_Tests___test_number_magnitude_key_dev_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_number_magnitude_key.dev.3)
- [test_query_sort_order_number.dev.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/13497/testReport/junit/alternator/test_query/Tests___Unit_Tests___test_query_sort_order_number_dev_1) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_query_sort_order_number.dev.1)
#### Build Details:
- Duration: 8 hr 4 min
- Builder: i-035c0c2ba32ad675a (m5ad.8xlarge)

michoecho commented:
> > > @xemul you requested my review of this patch. Not coincidentally I gave a reference to the Java implementation. I think our safest bet here is to copy Cassandra, which is most likely using java.math.BigDecimal.compareTo() (needs to be checked), and not use our own logic. cc @denesb
> > 
> > 
> > @kostja You mean the semantics, or the implementation? As far as I can see in the source you linked, the Java implementation is afflicted with the very problem this PR is trying to fix.
> 
> How did you arrive to this? AFAICS compareTo avoids inflation as much as it can.

@kostja Oh, yes, you're right, I misread it. I thought the `val.bigMultiplyPowerTen` can get called with an arbitrarily large argument and didn't realize that this is prevented by the early return above (`if (xae < yae)`). (Because if the difference in `.scale` is bigger than the difference in the order of magnitude of `intVal`, then the early return will occur. So the argument to `bigMultiplyPowerTen` is limited to the order of magnitude of `intVal`, which is just what we want).

> As I showed in the cqlpy test above, Cassandra does not have the user-facing bug that we have. Whether it uses the specific Java function you mentioned or something else to compare "decimal" clustering keys, I don't know.

@nyh Yes, but you showed it only for 4 specific values. With a black box test like this, you don't know whether your test proves that there is no problem, or it just doesn't enter the code path with the problem.

michoecho commented:
> QQ: what about values that ARE close but do expand to huge values and cause OOM? Would we still be exposed to the issue for these even with this PR, @denesb ?

No. If the values ARE close, then the issue doesn't occur even without the PR.

The comparison normalizes the values to the same exponent, by choosing the value with the greater exponent, lowering the exponent to match the other value and multiplying the significand accordingly. For example, to compare 123E102 and 23456E100, the left value is normalized to 12300E100, and then 12300 and 23456 are compared.

If the values are close, then — by the very definition of "close" — the smaller significand will happen to be "expanded" exactly to the length of of the greater significand.

So if the values are close, then the comparison algorithm allocates an amount of memory which is at most big as the size of arguments passed to it.

nyh commented:
> @nyh Yes, but you showed it only for 4 specific values. With a black box test like this, you don't know whether your test proves that there is no problem, or it just doesn't enter the code path with the problem.

Of course, but in this case we have a theory what caused a problem - comparing these different-scaled numbers - and since we know both Cassandra and Scylla need to compare these numbers to correctly sort the clustering keys, and since Scylla botches up those comparison (hangs in a very long and memory-consuming loop) and Cassandra doesn't, clearly we didn't copy this bug from them and their implementation of the comparison doesn't have this bug. It doesn't prove, of course, that it doesn't have other bugs. And it doesn't mean we need to copy Cassandra's or Java's implementation instead of writing our own - and most definitely it doesn't prove that if we copy one specific function that you **believe** (without proof) that it implements this comparison in Cassandra it will actually help.

nyh commented:
> #### 🔴 CI State: FAILURE

Your patch broke several existing Alternator tests (unsurprisingly,. test that check **decimal** values) and one boost CQL test.

denesb commented:
> Regarding tests, this looks like a good candidate for property testing. Generate two numbers, compare them using both the smart algorithm and a dumb-but-obvious algorithm, assert the result is the same. (Generating two numbers in a way that makes the test valuable is the hard part here, e.g. just randomly generating two small numbers won't provide much value)

We will have to generate small numbers in such a test, otherwise the test will take minutes and possibly crash.

denesb commented:
New in v3:
* Handle negative numbers
* Add tests with negative numbers
* Integrate the test from Nadav (but make the number much-much higher).

nyh commented:
> Regarding tests, this looks like a good candidate for property testing. Generate two numbers, compare them using both the smart algorithm and a dumb-but-obvious algorithm, assert the result is the same. (Generating two numbers in a way that makes the test valuable is the hard part here, e.g. just randomly generating two small numbers won't provide much value)

There's a saying that "Perfect is the enemy of good". Yes, we could strive to have **perfect** tests that could prove mathematically that it covers all the possible numbers. I'm sure there has been a lot of research on it (which I never followed). Sadly, it is much more likely that in practice we'll end up with something like some dtests that have 200 lines of code making up "random items" and things like that which end up _not_ covering any of the risky cases that the developer knows exists - and the developer doesn't even know that because nobody can follow these 200 lines of code to understand which scenarios are tested.

The "good", as opposed to "the perfect", is when the developer **who originally wrote the code** writes a test covering the edge cases that he understood - during his development - might be problematic. In our example, somebody wrote this original number-comparison code, and he knew that his implementation is all about converting two different exponents into one, so it was a good candidate to write tests with different exponents, like 1e-100 and 1e+100. And if this person thought a bit more he would have tested also 1e-10000000000 and 1e+10000000000000. Yes, the developer might have missed this idea to do this last test, which is why this testing technique is "good" but not "best". But even if it's not best, it's better than the absolutely-nothing we had today testing this feature.

nyh commented:
>     * Integrate the test from Nadav (but make the number much-much higher).

A small mathematical correction - the number you used is not "much-much higher" :-) In my example I used a number with 1000000000 digits. In your example, it has 1000000023 digits. Just 23 more digits (the "19866597869857659876855" you tacked on the front). So I guess it doesn't hurt to spice up this test with more digits of mantissa, but it doesn't make the size of the allocation in the old buggy algorithm significantly higher.

avikivity commented on utils/big_decimal.cc:
```diff
@@ -143,6 +143,99 @@ std::strong_ordering big_decimal::operator<=>(const big_decimal& other) const
+    }
+
+    const int sign = boost::multiprecision::sign(_unscaled_value) < 0 ? -1 : 1;
+    const int sign_other = boost::multiprecision::sign(other._unscaled_value) < 0 ? -1 : 1;
```
Why coerce sign into -1/+1? if both signs equal zero, we don't need to compare anything, and if one is zero and the other is nonzero, the comparison below works. With the coercion, you lose the fast path when comparing 0 and some large positive number.

denesb replied:
I couldn't find any docs on this, so I was afraid to make assumptions.

denesb replied:
I removed the coercion in v4.

avikivity commented on utils/big_decimal.cc:
```diff
@@ -143,6 +143,99 @@ std::strong_ordering big_decimal::operator<=>(const big_decimal& other) const
+    // Now we can rewrite the original numbers as follows:
+    //
+    //      auto number = _unscaled_value * std::pow(10, -_scale);
+    //      auto number_other_approx = _unscaled_value * std::pow(2, diff_scale_2) * std::pow(10, -other._scale);
```
Wouldn't it be other._unscaled_value?

denesb replied:
`_other._unscaled_value = _unscaled_value * std::pow(2, diff_scale_2)`.
We have expressed `_other._unscaled_value` in terms of `_usncaled_value`, so the two unscaled values are the same and we can directly compare the scales instead.

avikivity commented on utils/big_decimal.cc:
```diff
@@ -143,6 +143,99 @@ std::strong_ordering big_decimal::operator<=>(const big_decimal& other) const
+    //      auto number = _unscaled_value * std::pow(10, -_scale);
+    //      auto number_other_approx = _unscaled_value * std::pow(2, diff_scale_2) * std::pow(10, -other._scale);
+    //
+    // Notice that number_other_approx != number_other, but it is close, the following holds:
```
It would be easier to notice it if we had the formula for number_other to compare to.

denesb replied:
In v4 I added here the definition of `number_other` too, to make this easier to follow.

denesb commented:
> > ```
> > * Integrate the test from Nadav (but make the number much-much higher).
> > ```
> 
> A small mathematical correction - the number you used is not "much-much higher" :-) In my example I used a number with 1000000000 digits. In your example, it has 1000000023 digits. Just 23 more digits (the "19866597869857659876855" you tacked on the front). So I guess it doesn't hurt to spice up this test with more digits of mantissa, but it doesn't make the size of the allocation in the old buggy algorithm significantly higher.

Yes, I should have used the term "bigger" and not in the mathematical sense, just in the physical sense (the number as written takes up more space :)).

denesb commented on utils/big_decimal.cc:
```diff
@@ -143,6 +143,99 @@ std::strong_ordering big_decimal::operator<=>(const big_decimal& other) const
+    }
+
+    const int is_zero = !_unscaled_value;
+    const int is_zero_other = !other._unscaled_value;
```
See above. I checked and indeed boost seems to follow this convention. I added a test for it just in case.

denesb replied:
I dropped this zero comparison in v4, I added a check using `sign` instead.

nyh commented on test/cqlpy/test_decimal.py:
```diff
@@ -0,0 +1,46 @@
+# There are additional tests involving decimals in specific contexts in other
+# files - e.g., aggregating decimals in test_aggregate.py, casting decimals
+# in test_cast_data.py, and decimals in JSON in test_json.py
+#############################################################################
```
Please consider using for this file the name test_type_decimal.py, like I suggested, and not test_decimal. I have other files `test_type_*` (date, duration, string, time, timestamp, uuid). I think it's good to know that these tests are about the CQL type "decimal" and not other things which might be called "decimal" (for type_timestamp it's even more obvious that it's an important distinction).

denesb replied:
Done in v4.1.

nyh commented on test/boost/big_decimal_test.cc:
```diff
@@ -203,3 +203,112 @@ BOOST_AUTO_TEST_CASE(test_big_decimal_sub) {
+        }
+    };
+    // Some of the numbers in this test are truly humongous, converting them to string will stress the machine and will make the test output unusable.
+    // So avoid rendering the numbers using big_decimal::to_string(), just print the unscaled value and the scale instead.
```
This observation really points to a bug in bid_decimal::to_string() - there is no reason why `1789679678670986897698780986986E99999999` can't be printed like that, in scientific notation, and needs to be written as 100 million digits. This bug in to_string() probably causes https://github.com/scylladb/scylladb/issues/8002 and the fact we use it probably also explains https://github.com/scylladb/scylladb/issues/7972. But you don't need to solve it now.

nyh commented on test/boost/big_decimal_test.cc:
```diff
@@ -203,3 +203,112 @@ BOOST_AUTO_TEST_CASE(test_big_decimal_sub) {
+    test_cmp(big_decimal("-1.3345E3"), big_decimal("0"), std::strong_ordering::less);
+    test_cmp(big_decimal("0"), big_decimal("1"), std::strong_ordering::less);
+    test_cmp(big_decimal("0"), big_decimal("187687E3"), std::strong_ordering::less);
+    test_cmp(big_decimal("0"), big_decimal("187687E3"), std::strong_ordering::less);
```
The two last checks are identical, I assume you wanted to do something different?

denesb replied:
I think I wanted to have a test with a negative number too. I will adjust.

denesb replied:
Made the last one test with negatives in v5.

denesb commented:
New in v4.2: added a test case which fails with the sqrt vs log2 mistake (@michoecho).

nyh commented on test/boost/big_decimal_test.cc:
```diff
@@ -203,3 +203,116 @@ BOOST_AUTO_TEST_CASE(test_big_decimal_sub) {
+    // Test compare numbers with huge differences, this used to cause the comparison to allocate large amount of memory and generate stalls.
+    // Reproduces https://github.com/scylladb/scylladb/issues/21716
+    test_cmp(big_decimal("1789679678670986897698780986986E99999999"), big_decimal("1789679678670986897698780986986E-99999999"), std::strong_ordering::greater);
+    test_cmp(big_decimal(max_scale, 1), big_decimal("1"), std::strong_ordering::less);
```
I didn't understand the expected ordering until I saw the comment below "note scale has implicit inverted sign"... :-(

denesb replied:
Yes, it was very confusing to write this test. I kept using the wrong expected outcome.

nyh commented on test/boost/big_decimal_test.cc:
```diff
@@ -203,3 +203,116 @@ BOOST_AUTO_TEST_CASE(test_big_decimal_sub) {
+    test_cmp(big_decimal("1.0000000000000000000001"), big_decimal("1.00000000000000000000010000000000000007"), std::strong_ordering::less);
+    test_cmp(big_decimal("1.00000000000000000000010000000000000007"), big_decimal("1.0000000000000000000001"), std::strong_ordering::greater);
+    test_cmp(big_decimal("-1.0000000000000000000001"), big_decimal("-1.00000000000000000000010000000000000007"), std::strong_ordering::greater);
+    test_cmp(big_decimal("-1.00000000000000000000010000000000000007"), big_decimal("-1.0000000000000000000001"), std::strong_ordering::less);
```
You are missing here a test with very large scale but very close mantissa - for example
1e1000000 and 99e999998
This will convert both huge numbers to the same scale, but should be quick and won't suffer the original bug.

denesb replied:
Added a test in v5 (also with negative variant).

nyh commented on utils/big_decimal.cc:
```diff
@@ -143,6 +143,97 @@ std::strong_ordering big_decimal::operator<=>(const big_decimal& other) const
+    // To compare them, we make the unscaled values the same (or close), so we can directly compare the scales.
+    // To do that we want to compute diff_scale_2:
+    //
+    //      auto unscaled_diff = _unscaled_value / other._unscaled_value;
```
Nitpick: this isn't a "diff" (which is subtraction), but a ratio (division).

nyh replied:
Oh, I see in the next line you look at the logarithm, so the ratio really becomes a "diff". But at this point of the comment, it's still a ratio. Only the next line is a diff.

denesb replied:
Renamed to `unscaled_ratio` in v5.

nyh commented on utils/big_decimal.cc:
```diff
@@ -143,6 +143,97 @@ std::strong_ordering big_decimal::operator<=>(const big_decimal& other) const
+    // To do that we want to compute diff_scale_2:
+    //
+    //      auto unscaled_diff = _unscaled_value / other._unscaled_value;
+    //      auto diff_scale_2 = log2(unscaled_diff);
```
In the explanation above, "scale" was about base-10 logarithms, not base-2. Below you switch to log2 for efficiency, but I think this switch is premature in this line of the comment.

denesb replied:
Renamed to `unscaled_ratio_log2` in v5.

nyh commented on utils/big_decimal.cc:
```diff
@@ -143,6 +143,97 @@ std::strong_ordering big_decimal::operator<=>(const big_decimal& other) const
+    //
+    // To avoid using division and then calculating a log2(), we use the number
+    // of bits in both numbers do infer log2(unscaled_diff) directly.
+    const int64_t bit_count = boost::multiprecision::msb(comparing_negatives ? -_unscaled_value : _unscaled_value);
```
A thought (feel free to ignore): instead of this comparing_negatives thing, you could have swapped the two values **and** their sign - so comparing -17, -18 would change to comparing 18, 17. Then you won't need to have these comparing_negative checks.

nyh replied:
Also, I think the name "bit_count" is ambiguous. Wouldn't it be clearer to call it "msb" or "highest_bit"?

nyh replied:
Ok, sounds good.

michoecho replied:
> We can create copies, but that can be potentially expensive

In the case when the values are negative, you already do that though. (When passing the negated argument to `msb`).

denesb replied:
Good point, let me see how would the code look with swap.

denesb replied:
Done in v6. I had to add `operator-()` too.

nyh reviewed (approved):
I think the code is correct, but I'm not 100% confident I followed the math and didn't miss anything (at least I didn't see any gross mistakes, and if there were any gross mistakes the tests you wrote would have noticed them too).
I left a few trivial and optional nitpicks for your consideration.

denesb commented:
New in v5:
* Use better name for terms in the code and comments.
* Fix duplicate test case (make one of them test with negative).
* Add one more test for very close numbers.

denesb commented:
New in v7: Avoid recursion by moving the positive/nonzero/different-scale compare into its own method.

denesb commented on utils/big_decimal.cc:
```diff
@@ -143,6 +143,100 @@ std::strong_ordering big_decimal::operator<=>(const big_decimal& other) const
+    if (sign < 0) {
+        auto a = -*this;
+        auto b = -other;
+        return b <=> a;
```
Fixed in v7. I promoted the rest of this method into its own method.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13551/)
:white_check_mark: - [Unit Tests Custom](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13551/)
**The following new/updated tests ran 100 times for each mode:**
:small_blue_diamond: boost/big_decimal_test
:white_check_mark: - [Docker Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci-docker/476/)
:white_check_mark: - [dtest with topology changes](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-consistent-topology-changes/5111/)
:white_check_mark: - [dtest with tablets](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release-with-tablets/2002/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/13126/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/13551/)
#### Build Details:
- Duration: 4 hr 5 min
- Builder: spider1.cloudius-systems.com

avikivity commented on test/boost/big_decimal_test.cc:
```diff
@@ -203,3 +203,118 @@ BOOST_AUTO_TEST_CASE(test_big_decimal_sub) {
+    test_cmp(big_decimal("134252342E9"), big_decimal("100023424234E19"), std::strong_ordering::less);
+    test_cmp(big_decimal("-1.1221310000000023424234E9"), big_decimal("-1.267687687968768686"), std::strong_ordering::less);
+    test_cmp(big_decimal("-1.267687687968768686"), big_decimal("-1.1221310000000023424234E9"), std::strong_ordering::greater);
+    test_cmp(big_decimal("-134252342E9"), big_decimal("-100023424234E19"), std::strong_ordering::greater);
```
These test don't stress the area where the numbers are close, but the scales are very far.

Examples:

1e20 vs "9" * 20      (python syntax for repeating "9")
1e20 vs "1" "0"*19 "1"

avikivity replied:
I see below you have tests that to test it.
